common/models/transducer/alignment.py

```diff
@@ -1,16 +1,33 @@
 from returnn.tf.util.basic import get_shape_dim, check_input_dim
+from returnn.tf.util.data import Data
+from returnn.tf.layers.base import LayerBase
+from tensorflow import Tensor
+from typing import List, Union, Protocol
 
 
-def rna_alignment(source, **kwargs):
+class SourceType(Protocol):
   """
-  Used only to create alignments according to RNA loss function.
-  :sources: [output_log_prob, real_target, "base:encoder"]
-  :return: alignments: [B, T] for each frame a value in [0:blank_ix]
+  Defines the type of source functions returned by eval_layer.
+  """
+  def __call__(self, i: int,
+               auto_convert: bool = True,
+               enforce_batch_major: bool = False,
+               as_data: bool = False) -> Union[Data, Tensor]:
+    ...
```

Comment:
This is how I end up doing. We can maybe add a file called `common_types.py` somewhere to save complicated type definitions such as this.

Reply:
Is this the proper way to define such a function API?

But yes, this is not what I was suggestion. I was suggesting to do it like we already have it in other RETURNN code. Very simple, just specify it in the docstring what the function does.

I think this is too much overkill/overengineered.

Also, we really should avoid to copy & paste too much RETURNN stuff here (such as the API itself).

Reply:
I found it [here](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols].

> But yes, this is not what I was suggestion. I was suggesting to do it like we already have it in other RETURNN code. Very simple, just specify it in the docstring what the function does. I think this is too much overkill/overengineered. Also, we really should avoid to copy & paste too much RETURNN stuff here (such as the API itself).

I couldn't find any similiar case in RETURNN where a callbeck is used as param of a function. Could you give me a pointer? 
`setup()` is a subfunction and doesn't really have a Public API I noticed. It is not straight forward why it takes the parameters it take and what does it return.

Reply:
> I couldn't find any similiar case in RETURNN where a callbeck is used as param of a function. Could you give me a pointer?

[Here](https://returnn.readthedocs.io/en/latest/configuration_reference/pretraining.html), [here](https://github.com/rwth-i6/returnn/blob/master/CHANGELOG.md#2020-06-05-new-tf-dataset-pipeline-via-tfdataset-discussion), [here](https://github.com/rwth-i6/returnn/blob/22be8e029f4ea1f3fb7b319beaf99711176046b4/returnn/pretrain.py#L22), etc.

It's not really so important how exactly you do it. It should just be clear.

> `setup()` is a subfunction and doesn't really have a Public API I noticed. It is not straight forward why it takes the parameters it take and what does it return.

What `setup` do you mean?

Reply:
`source()` i meant sorry

Reply:
Ahh, i see. 
But then you get no help from typing with mypy.

Reply:
> But then you get no help from typing with mypy.

Yes. But this should improved on the RETURNN side directly. Also, we still want to keep Python 2 compatibility, at least in main RETURNN. Not here in returnn-experiments. And also we don't really want to over-engineer things.